Look up default strategy from Warden env on auth fail

alexsoble opened the pull request:
# What does this do? 

When a user attempts to login and fails, this sets the `:strategy` attribute of the LoginActivity record based on the default strategies being used instead of always setting to "database_authenticatable". 

See #11 for more background.

# How does it work? 

This relies on data in the `env` argument passed down by the `before_failure` Warden hook. Here's an example of the shape of `env['warden'].config` in the context of an education-related web app: 

```
# env['warden'].config:

{:default_scope=>:educator, 
  :scope_defaults=>{}, 
  :default_strategies=>{:educator=>[:ldap_authenticatable_tiny, :rememberable]}, 
  :intercept_401=>false, 
  :failure_app=>#<Devise::Delegator:0x007fd464db0840>}
```

# Edge cases

+ If the array of default strategies is empty for some reason, strategy will be set to `nil`, which seems right. 
+ @ankane, are there specific Warden-related edge cases you recommend testing against?

ankane commented on lib/auth_trail/manager.rb:
```diff
@@ -28,8 +28,12 @@ def before_failure(env, opts)
             identity = request.params[opts[:scope]] && request.params[opts[:scope]][:email] rescue nil
 
+            warden_config = env['warden'].config
+            default_scope = warden_config[:default_scope]
```
`opts` will have the scope used.

ankane commented on lib/auth_trail/manager.rb:
```diff
@@ -28,8 +28,12 @@ def before_failure(env, opts)
 
+            warden_config = env['warden'].config
+            default_scope = warden_config[:default_scope]
+            default_strategy = warden_config[:default_strategies][default_scope][0]
```
Let's take the first one that's not `:rememberable`, and cast it to a string.

alexsoble commented:
Good morning @ankane! I read in scope from opts and excluded `:rememberable` in 400c14fba351c5cdf5dbc0cf01de4b0b9add70f3. 

I tested the latest code on a local copy of the education web app (Student Insights) and it read the strategy correctly. Do you have suggestions for more robust ways to test?

ankane commented:
Thanks @alexsoble. When testing, I just realized `env["warden"]` is the `auth` object from `AuthTrail::Manager.after_set_user`, which contains the winning strategy, so we don't have to assume the default. I think we can use the same logic.

alexsoble commented:
Hey @ankane! Can you say a little more about what you mean when you're saying that `env['warden']` contains the winning strategy? 

To re-create an app with multiple strategies at work, I created a fake strategy locally. Here's what I see when I print out `env['warden']`:

```
Warden::Proxy:70242860777720 @config={
  :default_scope=>:educator, 
  :scope_defaults=>{}, 
  :default_strategies=>{:educator=>[:ldap_authenticatable_tiny, :fake_strategy, :rememberable]},
  :intercept_401=>false, 
  :failure_app=>#<Devise::Delegator:0x00007fc55e0399b0>
}
```

ankane commented:
The proxy object looks simple when you print it, but it actually has a lot of methods, including a `winning_strategy` one.

```ruby
p env["warden"].methods - Object.methods
```

alexsoble commented:
@ankane Oooooh, thanks! I'll explore that!

alexsoble commented:
@ankane I updated the PR to use the `winning_strategy` method. The method that displays a name string for each strategy is private ([devise/lib/devise/strategies/authenticatable.rb#L171](https://github.com/plataformatec/devise/blob/master/lib/devise/strategies/authenticatable.rb#L171)) so I copied over a bit of that class-to-string logic. Also updated the README.
